python/lsst/daf/butler/core/sqlDatabaseDict.py

```diff
@@ -0,0 +1,135 @@
+        may also be passed directly, though of course this not portable to
+        other `DatabaseDict` subclasses.
+    key : `str`
+        The name of the field to be used as the dictionary key.
+    value : `type` (`namedtuple`)
+        The type used for the dictionary's values, typically a `namedtuple`.
+        Must have a ``_field`` class attribute that is a tuple of field
+        names also appearing in ``fields``, and a `_make` attribute to
+        construct it from a sequence of values.
+    engine : `sqlalchemy.engine.Engine`
+        A SQLAlchemy connection object.  If not None, ``config["db"]`` is
+        ignored.
+    """
+
+    COLUMN_TYPES = {str: String, int: Integer, float: Float,
+                    bool: Boolean, bytes: LargeBinary, datetime: DateTime}
+
+    def __init__(self, config, fields, key, value, engine=None):
+        allColumns = []
+        for name, type_ in fields.items():
+            column = Column(name, self.COLUMN_TYPES.get(type_, type_), primary_key=(name == key))
+            allColumns.append(column)
+        if engine is None:
+            engine = create_engine(config['db'])
+        self._key = key
+        self._value = value
+        self._engine = engine
+        metadata = MetaData()
+        self._table = Table(config["table"], metadata, *allColumns)
+        metadata.create_all(self._engine)
+        valueColumns = [getattr(self._table.columns, name) for name in self._value._fields]
+        keyColumn = getattr(self._table.columns, key)
+        self._getSql = select(valueColumns).where(keyColumn == bindparam("key"))
+        self._updateSql = self._table.update().where(keyColumn == bindparam("key"))
+        self._delSql = self._table.delete().where(keyColumn == bindparam("key"))
+        self._keysSql = select([keyColumn])
+        self._valuesSql = select(valueColumns)
+        self._itemsSql = select([keyColumn] + valueColumns)
+        self._lenSql = select([func.count(keyColumn)])
+
+    def __getitem__(self, key):
+        with self._engine.begin() as connection:
+            row = connection.execute(self._getSql, key=key).fetchone()
+            if row is None:
+                raise KeyError("{} not found".format(key))
+            return self._value._make(row)
+
+    def __setitem__(self, key, value):
+        assert isinstance(value, self._value)
+        # Try insert first, as we expect that to be the most commmon usage
+        # pattern.
+        kwds = value._asdict()
+        with self._engine.begin() as connection:
+            try:
+                if self._key not in self._value._fields:
+                    kwds[self._key] = key
+                connection.execute(self._table.insert(), **kwds)
+                return
+            except IntegrityError:
+                pass
+        # If we fail due to an IntegrityError (i.e. duplicate primary key values),
+        # try to do an update instead.
+        kwds.discard(self._key)
```

Comment:
Is there a test for this? I think `kwds` is a dict (`OrderedDict` probably) and that does not have a `discard()` method. I think this should be `pop()` or `del`.